<?xml version="1.0" ?>

<robot name="eSAUl" xmlns:xacro="http://www.ros.org/wiki/xacro">
  
  <xacro:property name="chassis_mass" value="40" />
  <xacro:property name="chassis_length" value="0.45" />
  <xacro:property name="chassis_width" value="0.22" />
  <xacro:property name="chassis_height" value="0.13" />

  <xacro:property name="track_mass" value="0.3" />
  <xacro:property name="track_rad" value="0.03" />
  <xacro:property name="track_length" value="0.04" />

  <xacro:property name="box_length" value="0.35" />
  <xacro:property name="box_width" value="0.03" />
  <xacro:property name="box_height" value="0.06" />
  <xacro:property name="wheel_mass" value="10" />

  <xacro:property name="lidar_mass" value="0.1" />
  <xacro:property name="lidar_height" value="0.02"/>
  <xacro:property name="lidar_radius" value="0.04"/>

  <xacro:property name="wheel_friction" value="1.75" />

  <!-- INERTIAL MACROS -->
  <xacro:macro name="null_inertial">
    <inertial>
      <mass value="0.01"/>
      <inertia ixx="0.01" ixy="0" ixz="0"
               iyy="0.01" iyz="0"
               izz="0.01"/>
      </inertial>
  </xacro:macro>

  <xacro:macro name="solid_cuboid_inertial"
               params="width length height mass *origin">
    <inertial>
      <xacro:insert_block name="origin"/>
      <mass value="${mass}"/>
      <inertia ixx="${mass * (width * width + height * height) / 12}"
               ixy="0" ixz="0"
               iyy="${mass * (length * length + height * height) / 12}"
               iyz="0"
               izz="${mass * (width * width + length * length) / 12}"/>
    </inertial>
  </xacro:macro>

  <!-- <xacro:macro name="solid_cylinder_inertial"
               params="r h mass *origin">
    <inertial>
      <xacro:insert_block name="origin"/>
      <mass value="${mass}"/>
      <inertia ixx="${mass / 12 * (3 * r * r + h * h)}" ixy="0.0" ixz="0.0"
               iyy="${mass / 12 * (3 * r * r + h * h)}" iyz="0.0"
               izz="${mass / 2 * r * r}" />
    </inertial>
  </xacro:macro> -->

    <xacro:macro name="solid_cylinder_inertial"
               params="r h mass *origin">
    <inertial>
      <xacro:insert_block name="origin"/>
      <mass value="${mass}"/>
      <inertia ixx="${(mass * h * h)/12 + (mass * r * r)/4}" ixy="0.0" ixz="0.0"
               iyy="${(mass * h * h)/12 + (mass * r * r)/4}" iyz="0.0"
               izz="${mass / 2 * r * r}" />
    </inertial>
  </xacro:macro>


  <!-- INERTIAL MACROS END -->

  <link name="base_link">
    <collision name="collision_chassis">
      <!-- <origin rpy="0 0 0" xyz="0 0 ${track_rad}"/> -->
      <geometry>
        <box size="${chassis_length} ${chassis_width} ${chassis_height}"/>
      </geometry>
    </collision>
    <visual>
      <!-- <origin rpy="0 0 0" xyz="0 0 ${track_rad}"/> -->
      <geometry>
        <box size="${chassis_length} ${chassis_width} ${chassis_height}"/>
      </geometry>
    </visual>
    <xacro:solid_cuboid_inertial
          length="${chassis_length}" width="${chassis_width}"
          height="${chassis_height}" mass="${chassis_mass}">
      <origin xyz="0 0 0" rpy="0 0 0" />
    </xacro:solid_cuboid_inertial>
  </link>



  <xacro:macro name="link_track" params="name">
    <link name="${name}">
      <collision name="collision_left_track">
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <geometry>
          <box size="${box_length} ${box_width} ${box_height}"/>
        </geometry>
      </collision>
      <visual>
        <origin rpy="0 0 0" xyz="0 0 0"/>
        <geometry>
          <box size="${box_length} ${box_width} ${box_height}"/>
        </geometry>
        <material name="black"/>
      </visual>
      <xacro:solid_cuboid_inertial
            length="${box_length}" width="${box_width}"
            height="${box_height}" mass="${track_mass}">
        <origin xyz="0 0 0" rpy="0 0 0" />
      </xacro:solid_cuboid_inertial>
    </link>
    <gazebo reference="${name}">
      <mu1>0</mu1>
      <mu2>0</mu2>
      <material>Gazebo/Black</material>
    </gazebo>
  </xacro:macro>


  <xacro:macro name="link_wheel" params="name">
    <link name="${name}">
      <xacro:solid_cylinder_inertial
          r="${track_rad}" h="${track_length}"
          mass="${wheel_mass}" >
        <origin xyz="0 0 0" rpy="0 0 0" />
      </xacro:solid_cylinder_inertial>
      <collision name="link_right_wheel_collision">
        <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
        <geometry>
          <cylinder length="${track_length}" radius="${track_rad}"/>
        </geometry>
      </collision>
      <visual name="${name}_visual">
        <origin rpy="0 1.5707 1.5707" xyz="0 0 0"/>
        <geometry>
          <cylinder length="${track_length}" radius="${track_rad}"/>
        </geometry>
      </visual>
    </link>
    <gazebo reference="${name}">
      <mu1>1.5</mu1>
      <mu2>1.5</mu2>
      <material>Gazebo/Black</material>
      <kp>10000000</kp>
      <kd>1</kd>
    </gazebo>
  </xacro:macro>

  <xacro:macro name="joint_wheel_right" params="name child origin_xyz">
    <joint name="${name}" type="continuous">
      <origin rpy="0 0 0" xyz="${origin_xyz}"/>
      <child link="${child}"/>
      <parent link="link_right"/>
      <axis rpy="0 0 0" xyz="0 1 0"/>
      <limit effort="10000" velocity="1000"/>
      <!-- <use_parent_model_frame>1</use_parent_model_frame> -->
      <!-- <joint_properties damping="1.0" friction="1.0"/> -->
    </joint>
  </xacro:macro>

    <xacro:macro name="joint_wheel_left" params="name child origin_xyz">
    <joint name="${name}" type="continuous">
      <origin rpy="0 0 0" xyz="${origin_xyz}"/>
      <child link="${child}"/>
      <parent link="link_left"/>
      <axis rpy="0 0 0" xyz="0 1 0"/>
      <limit effort="10000" velocity="1000"/>
      <!-- <use_parent_model_frame>1</use_parent_model_frame> -->
      <!-- <joint_properties damping="1.0" friction="1.0"/> -->
    </joint>
  </xacro:macro>

  <xacro:macro name="joint_track" params="name child origin_xyz">
    <joint name="${name}" type="fixed">
      <origin rpy="0 0 0" xyz="${origin_xyz}"/>
      <child link="${child}"/>
      <parent link="base_link"/>
      <axis rpy="0 0 0" xyz="0 1 0"/>
      <joint_properties damping="1.0" friction="0.0"/>
    </joint>
  </xacro:macro>
<!-- 
  <link name="laser_frame">
    <xacro:solid_cylinder_inertial
        r="${lidar_radius}" h="${lidar_height}"
        mass="${lidar_mass}" >
      <origin xyz="0 0 0" rpy="0 0 0" />
    </xacro:solid_cylinder_inertial>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder length="${lidar_height}" radius="${lidar_radius}"/>
      </geometry>
    </collision>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <cylinder length="${lidar_height}" radius="${lidar_radius}"/>
      </geometry>
    </visual>
  </link>
  <gazebo reference="laser_frame">
    <material>Gazebo/Red</material>
  </gazebo>

  <joint name="laser_joint" type="fixed">
    <child link="laser_frame" />
    <parent link="base_link" />
    <origin xyz="0 0 ${chassis_height/2 + lidar_height/2}" />
  </joint>


 -->



  <!-- Hokuyo Laser -->
  <link name="lidar_link">
    <xacro:solid_cylinder_inertial
        r="${lidar_radius}" h="${lidar_height}"
        mass="${lidar_mass}" >
      <origin xyz="0 0 0" rpy="0 0 0" />
    </xacro:solid_cylinder_inertial>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder length="${lidar_height}" radius="${lidar_radius}"/>
      </geometry>
    </collision>

    <visual>
      <origin xyz="0 0 0" rpy="0 0 0"/>
      <geometry>
        <cylinder length="${lidar_height}" radius="${lidar_radius}"/>
      </geometry>
    </visual>
  </link>

  <gazebo reference="lidar_link">
    <material>Gazebo/Red</material>
    <sensor type="ray" name="head_hokuyo_sensor">
      <pose>0 0 0 0 0 0</pose>
      <visualize>false</visualize>
      <update_rate>40</update_rate>
      <ray>
        <scan>
          <horizontal>
            <samples>720</samples>
            <resolution>1</resolution>
            <min_angle>-3.14</min_angle>
            <max_angle>3.14</max_angle>
          </horizontal>
        </scan>
        <range>
          <min>0.10</min>
          <max>30.0</max>
          <resolution>0.01</resolution>
        </range>
        <noise>
          <type>gaussian</type>
          <!-- Noise parameters based on published spec for Hokuyo laser
               achieving "+-30mm" accuracy at range < 10m.  A mean of 0.0m and
               stddev of 0.01m will put 99.7% of samples within 0.03m of the true
               reading. -->
          <mean>0.0</mean>
          <stddev>0.01</stddev>
        </noise>
      </ray>
      <plugin name="gazebo_ros_head_hokuyo_controller" filename="libgazebo_ros_laser.so">
        <topicName>/eSAUl/laser/scan</topicName>
        <frameName>lidar_link</frameName>
      </plugin>
    </sensor>
  </gazebo>

  <joint name="lidar_joint" type="fixed">
    <axis xyz="0 1 0" />
    <origin xyz="0 0 ${chassis_height/2 + lidar_height/2}" rpy="0 0 0"/>
    <parent link="base_link"/>
    <child link="lidar_link"/>
  </joint>





  <xacro:link_track name="link_right" />
	<xacro:joint_track name="joint_right" child="link_right" origin_xyz="0 ${-chassis_width/2-box_width/1.5} ${-2*box_width}" />

  <!-- <xacro:link_wheel name="right_wheel_1" />
	<xacro:joint_wheel_right name="joint_right_wheel_1" child="right_wheel_1" origin_xyz="${box_length/2} 0 0" /> -->

  <xacro:link_wheel name="right_wheel_2" />
	<xacro:joint_wheel_right name="joint_right_wheel_2" child="right_wheel_2" origin_xyz="0 0 0" />
<!-- 
  <xacro:link_wheel name="right_wheel_3" />
	<xacro:joint_wheel_right name="joint_right_wheel_3" child="right_wheel_3" origin_xyz="${-box_length/2} 0 0" /> -->



  <xacro:link_track name="link_left" />
	<xacro:joint_track name="joint_left" child="link_left" origin_xyz="0 ${chassis_width/2+box_width/1.5} ${-2*box_width}" />
<!-- 
  <xacro:link_wheel name="left_wheel_1" />
	<xacro:joint_wheel_left name="joint_left_wheel_1" child="left_wheel_1" origin_xyz="${box_length/2} 0 0" /> -->

  <xacro:link_wheel name="left_wheel_2" />
	<xacro:joint_wheel_left name="joint_left_wheel_2" child="left_wheel_2" origin_xyz="0 0 0" />

  <!-- <xacro:link_wheel name="left_wheel_3" />
	<xacro:joint_wheel_left name="joint_left_wheel_3" child="left_wheel_3" origin_xyz="${-box_length/2} 0 0" /> -->


  	<gazebo>
    		<plugin name="ModelPlugin" filename="libesaul_description.so" >
      			<robotName>eSAUl</robotName>
            <joint_left1>joint_left_wheel_1</joint_left1>
            <joint_right1>joint_right_wheel_1</joint_right1>
            <joint_left2>joint_left_wheel_2</joint_left2>
            <joint_right2>joint_right_wheel_2</joint_right2>
            <joint_left3>joint_left_wheel_3</joint_left3>
            <joint_right3>joint_right_wheel_3</joint_right3>
            <width_chassis>${chassis_width}</width_chassis>
            <wh_rad>${track_rad}</wh_rad>
   		 </plugin>
	</gazebo>
  
 </robot>